Add support for AWS Cloud Map as service-discovery

kiranmeduri opened the pull request:
*Issue #, if available:*
#52 

*Description of changes:*
Add support for AWS Cloud Map as service-discovery

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

nckturner commented on pkg/controller/virtualnode.go:
```diff
@@ -296,3 +337,79 @@ func (c *Controller) handleVNodeMeshDeleting(ctx context.Context, vnode *appmesh
+
+	//It is okay to call Create multiple times for same service-name.
+	//It is also cheaper than calling get and then figuring out to create.
+	cloudmapService, err := c.cloud.CloudMapCreateService(ctx, cloudmapConfig, c.name)
```
Do we need to handle the case where the cloudmap namespace doesn't exist?  I guess its probably evident with the error returned from the cloudmap create service call, which is already logged.

kiranmeduri replied:
we probably should but creating namespace on the fly seems bad. CloudMap namespace should be created by admin, same can be said of mesh though. For now I will add a TODO to improve the experience.

lavignes commented on pkg/aws/cloudmap.go:
```diff
@@ -0,0 +1,319 @@
+	AttrK8sPod = "k8s.io/pod"
+	//AttrK8sNamespace is a custom attribute injected by app-mesh controller
+	AttrK8sNamespace = "k8s.io/namespace"
+)
```
https://docs.aws.amazon.com/AmazonECS/latest/developerguide/service-discovery.html

:+1: for REGION and AVAILABILITY_ZONE. This will be helpful for AZ-aware routing: aws/aws-app-mesh-roadmap#94

kiranmeduri replied:
Yes they are useful. Should we look into adding them as labels to pods and let register instance take over from there?

nckturner replied:
I can see those labels being useful for pods--how does the cluster admin decide which pods get them?

kiranmeduri replied:
For now we can lookup the node corresponding to pod and get that information. Below is the information from node. However, controller needs permissions to get node info. We can follow up this work in upcoming versions.

```
labels:
    alpha.eksctl.io/cluster-name: cluster-20190821
    alpha.eksctl.io/instance-id: i-08dd903c3cbe5db66
    alpha.eksctl.io/nodegroup-name: m5-l-1
    beta.kubernetes.io/arch: amd64
    beta.kubernetes.io/instance-type: m5.large
    beta.kubernetes.io/os: linux
    failure-domain.beta.kubernetes.io/region: us-west-2
    failure-domain.beta.kubernetes.io/zone: us-west-2a
    kubernetes.io/hostname: ip-192-168-118-179.us-west-2.compute.internal
    role: workers
```

nckturner commented on pkg/aws/cloudmap.go:
```diff
@@ -0,0 +1,351 @@
+		DnsConfig: &servicediscovery.DnsConfig{
+			NamespaceId:   awssdk.String(namespaceSummary.NamespaceID),
+			RoutingPolicy: awssdk.String(servicediscovery.RoutingPolicyMultivalue),
+			DnsRecords: []*servicediscovery.DnsRecord{
```
I think we need to add this conditionally, I am getting this error: 
```
E0903 15:44:25.848882       1 controller.go:443] error syncing 'appmesh-demo/colorteller-red': Error handling cloudmap service discovery for virtual node colorteller-red-appmesh-demo: InvalidInput: When you create a service using a namespace that has a type of HTTP, you can't include a DnsConfig element.
        status code: 400, request id: 8c76f6af-1c79-4aa7-a24a-a2b2c3de0e5b
```

kiranmeduri replied:
HTTP namespace will not work with AppMesh because those service-names cannot be resolved by VPC DNS. In the current setup, application container still resolves the name using DNS resolver which gives an endpoint that is ignored by Envoy and further resolved. https://github.com/envoyproxy/envoy/issues/6748 will help in avoiding DNS altogether in future. Until then we should be using PrivateDnsNamespace.

nckturner replied:
Ok we can call this out as its a little confusing.  Arguably, though, as long as the application gets some DNS response, then does everything work fine?  Whether its kube dns, route53, or some custom DNS solution the customer is running.

lavignes replied:
@nckturner
> as long as the application gets some DNS response, then does everything work fine
Yeah that is case right now. We're also looking into adding some work-around we can give customers since some people may not be allowed to create a PrivateDnsNamespace.

kiranmeduri replied:
Updated PR with better error message when provided namespace is non DNS_PRIVATE. Marked it with a TODO to support HTTP namespaces. That should unblock this PR.

kiranmeduri commented:
> I think there is a situation where things don't reconcile properly, or we are ignoring my namespace for some reason. I had example items created before updating to a controller built from this branch. I updated to the new controller, modified a virtual node to specify service discovery == cloudmap. The controller prints `cloudmap.go:273] No namespace found with name color-mesh`, and then even after creating the cloudmap namespace, and redeploying the controller, I'm seeing the same error.
> 
> **Update:** this was just me not having the correct cloudmap permissions on my nodes, which I realized after far too much debugging.

I will fix the error handling. We should also update documentation so that users do not trip on this. Also worth to look into https://docs.aws.amazon.com/eks/latest/userguide/iam-roles-for-service-accounts.html
